src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,393 @@
+import { ShaderGraphNode, PORT_TYPE_IN, PORT_TYPE_SWITCH, comp2Type } from './shader-graph-node.js';
+
+import { Texture } from '../../graphics/texture.js';
+
+import { PIXELFORMAT_R8_G8_B8_A8, TEXTURETYPE_RGBM } from '../../graphics/graphics.js';
+
+var id = 0;
+
+var _placeholderTex;
+
+/**
+ * @private
+ * @class
+ * @name pc.ShaderGraphBuilder
+ * @classdesc A Shader Graph Builder class.
+ * @param {pc.Application} app - Application with shader graph core nodes registered that will be used to
+ * build the shader graph.
+ */
+var ShaderGraphBuilder = function (app) {
+    id++;
+
+    this.app = app;
+
+    this._graph = new ShaderGraphNode();
+    this._graph.name = 'graphRoot_' + id;
+
+    // for validation
+    this._addedNamedInputs = {};
+    this._addedNodes = [];
+    this._addedOutputs = {};

```

Comment:
curious why these names start with 'added'?
Maybe just _namedInputs, _nodes, _outputs?

Reply:
I guess they can be renamed - they are just private data used for validation - maybe they can be tidied up by putting them into a _validationData struct?

Reply:
good idea.